File: src/carma_clock.py
import threading
import time

class CarmaClock:
    def __init__(self, simulation_mode=False):
        self._current_time = 0
        self._is_simulation_mode = simulation_mode
        self._is_initialized = not simulation_mode
        self._initialization_mutex = threading.Lock()
        self._initialization_cv = threading.Condition(lock=self._initialization_mutex)
        self._sleep_mutex = threading.Lock()
        self._sleep_holder = []

    def now_in_seconds(self) -> int:
        if self._is_simulation_mode:
            if not self._is_initialized:
                raise ValueError("Clock is not initialized!")
            # Round down as same behavior as chrono::duration_cast
            return int(self._current_time / 1000)
        else:
            return int(time.time())

    def now_in_milliseconds(self) -> int:
        if self._is_simulation_mode:
            if not self._is_initialized:
                raise ValueError("Clock is not initialized!")
            return int(self._current_time)
        else:
            return int(time.time() * 1000)

    def update(self, current_time: int) -> None:
        if not self._is_simulation_mode:
            raise ValueError("Clock is not in simulation mode!")

        self._current_time = current_time

        if not self._is_initialized:
            self._is_initialized = True
            with self._initialization_mutex:
                self._initialization_cv.notify_all()

        with self._sleep_mutex:
            self._process_sleeping_threads()

    def is_simulation_mode(self) -> bool:
        return self._is_simulation_mode

    def wait_for_initialization(self) -> None:
        if not self._is_initialized:
            with self._initialization_mutex:
                self._initialization_cv.wait_for(lambda: self._is_initialized)

    def sleep_until(self, future_time: int) -> None:
        if not self._is_initialized:
            raise ValueError("Clock is not initialized!")

        if self._is_simulation_mode:
            if future_time > self._current_time:
                sleep_cv_pair = threading.Condition(), threading.Lock()
                with self._sleep_mutex:
                    self._sleep_holder.append((future_time, sleep_cv_pair))
                               
                with sleep_cv_pair[1]:
                    with sleep_cv_pair[0]:
                        sleep_cv_pair[0].wait()

        else:
            time_to_sleep = max(0, (future_time - self.now_in_milliseconds()) / 1000)
            time.sleep(time_to_sleep)

    def sleep_for(self, time_to_sleep: int) -> None:
        self.sleep_until(self._current_time + time_to_sleep)

    def _process_sleeping_threads(self) -> None:
        current_time = self._current_time

        self._sleep_holder = [
            (time, cv_pair)
            for time, cv_pair in self._sleep_holder
            if current_time >= time
        ]

        for _, cv_pair in self._sleep_holder:
            with cv_pair[1]:
                cv_pair[0].notify()

# SYSTEM_SLEEP_TIME = 150

# # Create a sim clock
# clock = CarmaClock(simulation_mode=True)

# # Update the clock
# clock.update(1)

# # Thread function to update clock after sleeping
# def update_clock_after_sleep():
#     time.sleep(SYSTEM_SLEEP_TIME / 1000)  # Convert milliseconds to seconds
#     clock.update(2)

# # Create a thread to update clock after sleeping
# t = threading.Thread(target=update_clock_after_sleep)

# # Wait for initialization
# clock.wait_for_initialization()

# # Check if sim time matches expected value
# sim_time_now = clock.now_in_milliseconds()
# assert sim_time_now == 1, f"Sim time is {sim_time_now} instead of 1"

# # Measure start time
# start = time.time()

# # Sleep until sim time + 1
# clock.sleep_until((sim_time_now + 1))

# # Measure end time
# after = time.time()

# # Calculate elapsed time in milliseconds
# ms_count = int((after - start) * 1000)

# # Join the thread
# t.start()
# t.join()

# # Check if elapsed time matches expected sleep time
# assert abs(SYSTEM_SLEEP_TIME - ms_count) <= 5, f"Elapsed time: {ms_count}ms, Expected: {SYSTEM_SLEEP_TIME}ms"

File: src/carma_clock.cpp
/*
 * Copyright (C) 2023 LEIDOS.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#include "carma_clock.h"

#include <thread>
#include <exception>

namespace fwha_stol::lib::time {

CarmaClock::CarmaClock(bool simulation_mode) : _is_simulation_mode(simulation_mode) {
    // ready to go if not sim mode
    _is_initialized = ! _is_simulation_mode;
}
// // Copy constructor
CarmaClock::CarmaClock(const CarmaClock& other) 
    : _current_time(other._current_time),
      _is_simulation_mode(other._is_simulation_mode),
      _is_initialized(other._is_initialized),
      _sleep_holder(other._sleep_holder) {
    // Copy mutexes and condition variables
    // No direct copy is allowed for mutexes and condition variables, 
    // so we need to create new ones and leave the original ones in a
    // valid but unspecified state.
    // We'll just ensure that the new object has its own unique mutexes
    // and condition variables.
    // This will allow the copied object to function properly, but it won't
    // have the same threads blocked on it as the original.
    // for (const auto& sleep_entry : _sleep_holder) {
    //     std::mutex mutex;
    //     std::condition_variable cv;
    //     // We are not copying threads blocked on these mutexes and condition variables.
    //     // So, the copied object will have empty mutexes and condition variables.
    //     _sleep_holder.emplace_back(sleep_entry.first, std::move(mutex), std::move(cv));
    // }
}


timeStampMilliseconds CarmaClock::nowInMilliseconds() const {
    if (_is_simulation_mode) {
        if (!_is_initialized) {
            throw std::invalid_argument("Clock is not initialized!");
        }
        return _current_time;
    } else {
        using namespace std::chrono;
        return duration_cast< milliseconds>(system_clock::now().time_since_epoch()).count();
    }
}

timeStampSeconds CarmaClock::nowInSeconds() const {
    if (_is_simulation_mode) {
        if (!_is_initialized) {
            throw std::invalid_argument("Clock is not initialized!");
        }
        // purposefully round down as this is the same behavior as chrono::duration_cast
        return (_current_time / 1000);
    } else {
        using namespace std::chrono;
        return duration_cast< seconds>(system_clock::now().time_since_epoch()).count();
    }
}

void CarmaClock::update(timeStampMilliseconds current_time) {
    // check for sim time and throw exception if not in sim
    if (!_is_simulation_mode) {
        throw std::invalid_argument("Clock is not in simulation mode!");
    }
    _current_time = current_time;
    if (!_is_initialized) {
        // if not initialized then do it and let anyone waiting know
        _is_initialized = true;
        std::unique_lock lock(_initialization_mutex);
        // Notify all blocked threads
        _initialization_cv.notify_all();
    }
    // check to see if any sleeping threads need to be woken
    std::unique_lock lk(_sleep_mutex);
    for (auto iter = _sleep_holder.begin(); iter != _sleep_holder.end(); )
    {
        if(_current_time >= iter->first) {
            // found one, notify and then remove the record
            std::unique_lock lock(*iter->second.second);
            iter->second.first->notify_one();
            iter = _sleep_holder.erase(iter);
        } else {
            iter++;
        }
    }    
}

void CarmaClock::wait_for_initialization() {
    if (!_is_initialized) {
        std::unique_lock lock(_initialization_mutex);
        _initialization_cv.wait(lock, [this] { return _is_initialized; });
    }
}

void CarmaClock::sleep_for(timeStampMilliseconds time_to_sleep) {
    sleep_until( _current_time + time_to_sleep);
}

void CarmaClock::sleep_until(timeStampMilliseconds future_time) {
    if (!_is_initialized) {
        throw std::invalid_argument("Clock is not initialized!");
    }
    if (_is_simulation_mode) {
        // determine if we are at the time or not and skip sleep if past
        if (future_time > _current_time) {
            // create the CV and mutex to use
            sleepCVPair sleepCVPairValue = std::make_pair(
                std::make_shared<std::condition_variable>(),
                std::make_shared<std::mutex>()
            );
            {
                // add the time and the values to our list
                std::unique_lock lk(_sleep_mutex);
                _sleep_holder.emplace_back(future_time, sleepCVPairValue);
            }
            // wait for something to notify that this thread should proceed
            std::unique_lock lock(*sleepCVPairValue.second);
            sleepCVPairValue.first->wait(lock);
        }
    } else {
        // do system sleep
        using namespace std::chrono;
        std::chrono::system_clock::time_point futureTimePoint{std::chrono::milliseconds{future_time}};
        std::this_thread::sleep_until(futureTimePoint);
    }
}

}
